drivers/counter/counter_nrfx_timer.c

```diff
@@ -55,112 +65,173 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+		return (val - old) & top;
+	}
+
+	/* if top is not 2^n-1 */
+	return (val > old) ? (val - old) : val + top + 1 - old;
+
+}
+
+static void set_cc_int_pending(struct device *dev, u8_t chan)
+{
+	get_dev_data(dev)->cc_int_pending |= BIT(chan);
+	NRFX_IRQ_PENDING_SET(NRFX_IRQ_NUMBER_GET(get_nrfx_config(dev)->timer));
+}
+
+static int set_cc(struct device *dev, u8_t chan, u32_t val, u32_t flags)
+{
+	bool absolute = flags & COUNTER_ALARM_CFG_ABSOLUTE;
+	bool trigger_late = flags & COUNTER_ALARM_CFG_EXPIRE_WHEN_LATE;
+	NRF_TIMER_Type  *reg = get_nrfx_config(dev)->timer;
+	nrf_timer_event_t evt = CH2EVT(chan);
+	u32_t top = counter_get_top_value(dev);
+	int err = 0;
+	u32_t prev_val;
+	u32_t now;
+	u32_t diff;
+	u32_t max_abs_val;
+
+	max_abs_val = top - get_dev_data(dev)->guard_period;
+	/* First take care of a risk of an event coming from CC being set to
+	 * next tick. Reconfigure CC to future (now tick is the furtherest
+	 * future).
+	 */
+	now = read(dev);
+	prev_val = nrf_timer_cc_read(reg, chan);
+	nrf_timer_cc_write(reg, chan, now);
+	nrf_timer_event_clear(reg, evt);
+
+	if (!absolute) {
+		trigger_late = val < (top / 2);
+		/* limit max to detect short relative being set too late. */
+		max_abs_val >>= trigger_late;
+		val = ticks_add(now, val, top);
+	}
+
+	nrf_timer_cc_write(reg, chan, val);
+	diff = ticks_sub(val - 1, read(dev), top);
+	if (diff > max_abs_val) {
+		if (absolute) {
+			err = -ETIME;
+		}
 
-	if (remainder > ticks) {
-		cc_val += ticks;
+		if (trigger_late) {
+			set_cc_int_pending(dev, chan);
+		} else {
+			get_nrfx_config(dev)->ch_data[chan].callback = NULL;
+		}
 	} else {
-		cc_val = ticks - remainder;
+		nrf_timer_int_enable(reg, nrf_timer_compare_int_get(chan));
 	}
 
-	return cc_val;
+	return err;
 }
 
-static int counter_nrfx_set_alarm(struct device *dev, u8_t chan_id,
-				  const struct counter_alarm_cfg *alarm_cfg)
+static int set_alarm(struct device *dev, u8_t chan,
+			const struct counter_alarm_cfg *alarm_cfg)
 {
 	const struct counter_nrfx_config *nrfx_config = get_nrfx_config(dev);
-	const nrfx_timer_t *timer = &nrfx_config->timer;
-	u32_t cc_val;
+	struct counter_nrfx_ch_data *chdata = &nrfx_config->ch_data[chan];
+	int err;
 
-	if (alarm_cfg->ticks > nrfx_timer_capture_get(timer, TOP_CH)) {
+	if (alarm_cfg->ticks >  get_top_value(dev)) {
 		return -EINVAL;
 	}
 
-	if (nrfx_config->ch_data[chan_id].callback) {
+	if (chdata->callback) {
 		return -EBUSY;
 	}
 
-	cc_val = counter_nrfx_get_cc_value(dev, alarm_cfg);
+	chdata->callback = alarm_cfg->callback;
+	chdata->user_data = alarm_cfg->user_data;
 
-	nrfx_config->ch_data[chan_id].callback = alarm_cfg->callback;
-	nrfx_config->ch_data[chan_id].user_data = alarm_cfg->user_data;
-
-	nrfx_timer_compare(timer, ID_TO_CC(chan_id), cc_val, true);
+	err = set_cc(dev, ID_TO_CC(chan), alarm_cfg->ticks, alarm_cfg->flags);
+	if (err != 0) {
+		chdata->callback = NULL;
+	}
 
-	return 0;
+	return err;
 }
 
 static void disable(struct device *dev, u8_t id)
```

Comment:
Why separate function, when it is called only by `cancel_alarm` that differs solely by returning `0`?

Reply:
merged to `cancel_alarm`